pandas/_libs/parsers.pyx

```diff
@@ -15,6 +15,13 @@ import warnings
 
 from pandas.util._exceptions import find_stack_level
 
+from pandas import StringDtype
+from pandas.core.arrays import (
```

Comment:
is there a viable way to do this outside of the cython code?

Reply:
A possible option might be to change `_maybe_upcast` to optionally return a values array + mask array, and then do the actual construction in python?

Although since `_maybe_upcast` is only being called in this file (and thus from cython), that won't help. Unless if we would propagate such a (values, mask) tuple (instead of ArrayLike) through the different calls and return that from the `TextReader.read` method.

Reply:
yah not worth contorting ourselves to avoid non-cython imports.  just if there's a convenient alternative

Reply:
Could be done yes, but would make logic more complex

Reply:
yah, never mind then.  thanks for taking a look